provider/win_driver_utils.py

```diff
@@ -194,30 +194,133 @@ def copy_file_to_samepath(session, test, params):
+    """
     device_name = driver_info_dict[driver_name]["device_name"]
     device_hwid = driver_info_dict[driver_name]["hwid"]
+
+    session = vm.wait_for_login()
+    output = _pnpdrv_info(session, device_name, ["DeviceID"])
+    # workaround for viostor/vioscsi to get data device id
+    device_id = output[0]
+    device_id = '&'.join(device_id.split('&amp;'))
+    find_devices = False
+    for hwid in device_hwid.split():
+        hwid = hwid.split('"')[1]
+        if hwid in device_id:
+            find_devices = True
+    if not find_devices:
+        test.fail("Didn't find driver info from guest %s" % output)
+    return device_id
+
+
+def load_driver(vm, test, params, load_method='enable'):
+    """
+    Load driver.
+
+    :param vm: vm object
+    :param test: kvm test object
+    :param params: the dict used for parameters
+    :param load_method: Load driver method
+    """
+    session = vm.wait_for_login()
+    driver_name = params["driver_name"]
+    if load_method != 'enable':
+        devcon_path = params["devcon_path"]
+        media_type = params.get("virtio_win_media_type", "iso")
+
+        device_hwid = driver_info_dict[driver_name]["hwid"]
+        return install_driver_by_virtio_media(session, test, devcon_path,
+                                              media_type, driver_name,
+                                              device_hwid)
+    else:
+        cmd = params["driver_enable_cmd"]
+        device_id = get_device_id(vm, test, driver_name)
+        return enable_driver(vm, test, cmd, device_id)
+
+
+def unload_driver(vm, test, params, load_method='enable'):
+    """
+    Unload driver.
+
+    :param vm: vm object
+    :param test: kvm test object
+    :param params: the dict used for parameters
+    :param load_method: Load driver method
+    """
     session = vm.wait_for_login()
-    uninstall_driver(session, test, devcon_path, driver_name,
-                     device_name, device_hwid)
+    driver_name = params["driver_name"]
+    if load_method != 'enable':
+        devcon_path = params["devcon_path"]
+
+        device_name = driver_info_dict[driver_name]["device_name"]
+        device_hwid = driver_info_dict[driver_name]["hwid"]
+        return uninstall_driver(session, test, devcon_path, driver_name,
+                                device_name, device_hwid)
+    else:
+        cmd = params["driver_disable_cmd"]
+        device_id = get_device_id(vm, test, driver_name)
+        return disable_driver(vm, test, cmd, device_id)
+
+
+def memory_leak_check(vm, test, params, load_method='enable'):
+    """
+    In order to let the driver verifier to catch memory leaks, driver
+    should be unloaded after driver function.Note that if want to use
+    this function, driver verifier should be enabled before driver function.
+
+    :param vm: vm object
+    :param test: kvm test object
+    :param params: the dict used for parameters
+    """
+    unload_driver(vm, test, params, load_method)
```

Comment:
Or, we can start a session here, and then transfer it to each sub-function, there we will not start a session in each sub-function.  @vivianQizhu What do you think?

Reply:
@xiagao That's a good idea, will be nice when you debug from the session logs. And then we might pass session instead of vm.

Reply:
@vivianQizhu  @xiagao  Thanks all, could you please help to check it again?